binary_transparency/firmware/cmd/hacker/modify_bundle/modify_bundle.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// modify_bundle is a hacker tool for modifying proof bundles.
+//
+// It can update the expected measurement and firmware image hashes of the specified
+// proof bundle file, optionally sign the firmware statement, and finally write the bundle
+// to the specified output file.
+//
+// Usage:
+//   go run ./cmd/hacker/modify_bundle/ --logtostderr --device=[dummy,usbarmory] --binary=/path/to/new/firmware/image --input=/path/to/bundle.json --output=/path/to/bundle.json
+//
+// The first time you use this tool there will be no prior firmware metadata
+// stored on the device and the tool will fail.  In this case, use the --force
+// flag to apply the update anyway thereby creating the metadata.
+// Subsequent invocations should then work without needing the --force flag.package main
+package main
+
+import (
+	"crypto/sha512"
+	"encoding/json"
+	"flag"
+	"io/ioutil"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/binary_transparency/firmware/api"
+	dummy_common "github.com/google/trillian-examples/binary_transparency/firmware/devices/dummy/common"
+	"github.com/google/trillian-examples/binary_transparency/firmware/devices/usbarmory"
+	"github.com/google/trillian-examples/binary_transparency/firmware/internal/crypto"
+)
+
+var (
+	input      = flag.String("input", "", "File path read input ProofBundle from.")
+	output     = flag.String("output", "", "File path to write output ProofBundle to.")
+	sign       = flag.Bool("sign", true, "Whether to use stolen key to sign manifest.")
+	binaryPath = flag.String("binary", "", "Replacement binary image.")
+	deviceID   = flag.String("device", "", "One of [dummy,usbarmory].")
+)
+
+func main() {
+	flag.Parse()
+
+	bundleRaw, err := ioutil.ReadFile(*input)
+	if err != nil {
+		glog.Exitf("Failed to read transparency bundle %q: %q", *input, err)
+	}
+	var pb api.ProofBundle
+	if err := json.Unmarshal(bundleRaw, &pb); err != nil {
+		glog.Exitf("Failed to parse proof bundle file: %q", err)
+	}
+	var fs api.FirmwareStatement
+	if err := json.Unmarshal(pb.ManifestStatement, &fs); err != nil {
+		glog.Exitf("Failed to parse FirmwareStatement: %q", err)
+	}
+	var fm api.FirmwareMetadata
+	if err := json.Unmarshal(fs.Metadata, &fm); err != nil {
+		glog.Exitf("Failed to parse FirmwareMetadata: %q", err)
+	}
+
+	if len(*binaryPath) > 0 {
```

Comment:
I wonder if it's worth erroring, or at least printing a warning if this isn't true. Without executing this block all we're doing is round tripping the input and serializing it again? Being noisier in this case makes it harder to use this tool wrong.

Reply:
That's a reasonable thought, and if this tool grows in functionality then we can wind that back if necessary.